designer/src/components/TemplateBuilder.tsx

```diff
@@ -0,0 +1,763 @@
+import {
+  Add as AddIcon,
+  Code as CodeIcon,
+  Delete as DeleteIcon,
+  Error as ErrorIcon,
+  Preview as PreviewIcon,
+  ArrowDropUpRounded as ArrowDropUpRoundedIcon,
+  ArrowDropDownRounded as ArrowDropDownRoundedIcon,
+} from '@mui/icons-material';
+import {
+  Alert,
+  Box,
+  Button,
+  Card,
+  CardContent,
+  CardHeader,
+  Dialog,
+  DialogActions,
+  DialogContent,
+  DialogTitle,
+  FormControl,
+  Grid,
+  IconButton,
+  InputLabel,
+  MenuItem,
+  Paper,
+  Select,
+  Tab,
+  Tabs,
+  TextField,
+  Tooltip,
+  Typography,
+} from '@mui/material';
+import Mustache from 'mustache';
+import React, {useCallback, useEffect, useMemo, useState} from 'react';
+
+/**
+ * Supported block types in the template builder
+ */
+type BlockType = 'text' | 'variable' | 'if' | 'unless';
+
+/**
+ * Represents a single building block in the template
+ */
+interface TemplateBlock {
+  id: string;
+  type: BlockType;
+  content: string;
+  children?: TemplateBlock[];
+}
+
+/**
+ * Represents a variable that can be used in the template
+ */
+interface Variable {
+  name: string;
+  displayName: string;
+  type: 'field' | 'system';
+  value?: string; // For preview purposes
+}
+
+/**
+ * Props for the MustacheTemplateBuilder component
+ */
+interface MustacheBuilderProps {
+  /** Initial template string to populate the builder */
+  initialTemplate?: string;
+  /** Available field variables */
+  variables: Variable[];
+  /** Available system variables */
+  systemVariables: Variable[];
+  /** Callback fired when template is saved */
+  onSave: (template: string) => void;
+  /** Controls dialog visibility */
+  open: boolean;
+  /** Callback fired when dialog is closed */
+  onClose: () => void;
+}
+
+/**
+ * Safely parses a Mustache template string into TemplateBlock format
+ */
+const parseTemplate = (
+  template: string
+): {blocks: TemplateBlock[]; error: string | null} => {
+  try {
+    const tokens = Mustache.parse(template);
+    const blocks = tokens
+      .map(convertMustacheTokenToBlock)
+      .filter(isTemplateBlock);
+    return {blocks, error: null};
+  } catch (error) {
+    return {
+      blocks: [
+        {
+          id: Math.random().toString(36).substring(2, 9),
+          type: 'text',
+          content: template || '',
+        },
+      ],
+      error: error instanceof Error ? error.message : 'Invalid template syntax',
+    };
+  }
+};
+
+/**
+ * Converts a Mustache AST token into our TemplateBlock format
+ */
+const convertMustacheTokenToBlock = (token: any): TemplateBlock | null => {
+  const [tokenType, content, ...rest] = token;
+
+  try {
+    switch (tokenType) {
+      case 'text':
+        return {
+          id: Math.random().toString(36).substring(2, 9),
```

Comment:
This is not guaranteed to be unique, what happens when you get a collision? Since you use this code a lot it should at least be a fn. Suggest just using an incrementing variable to get a unique id.